src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackBinary.cs

```diff
@@ -0,0 +1,1243 @@
+﻿// Copyright (c) All contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using MessagePack.Internal;
+
+namespace MessagePack
+{
+#if MESSAGEPACK_INTERNAL
+    internal
+#else
+    public
+#endif
+    static class MessagePackBinary
+    {
+        /// <summary>
+        /// Ties to write a <see cref="MessagePackCode.Nil"/> value.
+        /// </summary>
+        /// <param name="span">The memory region to write to.</param>
+        /// <returns>The number of bytes written or the binary negation of the number of bytes that would be written.</returns>
+        public static int WriteNil(Span<byte> span)
+        {
+            if (span.Length >= 1)
+            {
+                span[0] = MessagePackCode.Nil;
+                return 1;
+            }
+            else
+            {
+                return ~1;
```

Comment:
This is a pretty odd-looking pattern. Why not just return the negative value?

Reply:
Actually I borrowed the idea of returning the bitwise complement from Microsoft's BinarySearch method https://docs.microsoft.com/en-us/dotnet/api/system.array.binarysearch?view=netframework-4.7.2 while I think it should not make any difference if the negative value is returned (it does in BinarySearch because they need to return '0' as well as a value, but MessagePack doesn't have this requirement).

The main reasoning behind this code is that given this piece of code:
``` CSharp
public void WriteNil()
{
	int result = MessagePackBinary.WriteNil(this.writer.Span);
	if (result < 0)
	{
		result = MessagePackBinary.WriteNil(this.writer.GetSpan(~result));
	}

	this.writer.Advance(result);
}
```

It should be inlined in:
``` CSharp
public void WriteNil()
{
	int result;
	// Start code taken from WriteNil
	if (this.writer.Span.Length >= 1)
	{
		this.writer.Span[0] = MessagePackCode.Nil;
		result = 1;
	}
	else
	{
		result = ~1;
	}
	// End code taken from WriteNil
	
	if (result < 0) {
		int sizeHint = ~result;
		Span<byte> span;
		// Start code taken from GetSpan
		// Start code taken from Ensure
		if (this.writer.Span.Length < sizeHint)
		{
			EnsureMore(sizeHint);
		}
		// End code taken from Ensure
		span = this.writer.Span;
		// End code taken from GetSpan
		
		// Start code taken from WriteNil
		if (span.Length >= 1)
		{
			span[0] = MessagePackCode.Nil;
			result = 1;
		}
		else
		{
			result = ~1;
		}
		// End code taken from WriteNil
	}

	this.writer.Advance(result);
}
```
A clever enough optimizer should be able to:
- recognize sizeHint is a constant;
- recognize that the second write to span is exactly the same code as the first one (same branches, same check on the same variable);
- remove bound-checking where not needed.

So the code should be optimized to (propagation of constant values and bound checking on Span.Length):

``` CSharp
public void WriteNil()
{
	int result;
	// Start code taken from WriteNil
	if (this.writer.Span.Length >= 1)
	{
		this.writer.Span[0] = MessagePackCode.Nil;
		result = 1;
	}
	else
	{
		EnsureMore(1);
		// Start code taken from WriteNil
		if (this.writer.Span.Length >= 1)
		{
			this.writer.Span[0] = MessagePackCode.Nil;
			result = 1;
		}
		else
		{
			result = ~1;
		}
		// End code taken from WriteNil
	}
	// End code taken from WriteNil
	this.writer.Advance(result);
}
```
This code can be easily re-arranged in:

``` CSharp
public void WriteNil()
{
	int result;
	
	if (this.writer.Span.Length < 1)
	{
		EnsureMore(1); 
	}
	// If the JIT compiler was able to infer that after this call this.writer.Span.Length 
	// is always > 1, then this check could be removed as well and the call to 
	// this.writer.Advance(~1) - but I don't think it will. 
	// Still, in this code, the compile won't emit any bound-checking logic, while 
	// in the other one it should.
	if (this.writer.Span.Length >= 1) {
		this.writer.Span[0] = MessagePackCode.Nil;
		result = 1;
		this.writer.Advance(1);
	}
	else {
		this.writer.Advance(~1);
	}
}
``` 
Which is pretty much the same code it as now.

I tried on similar code both RyuJIT and Roslyn and they both support these kind of optimization.